src/Microsoft.AspNetCore.Routing/Builder/EndpointRoutingApplicationBuilderExtensions.cs

```diff
@@ -18,20 +19,39 @@ public static IApplicationBuilder UseEndpointRouting(this IApplicationBuilder bu
             return builder.UseEndpointRouting(null);
         }
 
-        public static IApplicationBuilder UseEndpointRouting(this IApplicationBuilder builder, Action<EndpointDataSourceBuilder> configure)
+        public static IApplicationBuilder UseEndpointRouting(this IApplicationBuilder builder, Action<IEndpointRouteBuilder> configure)
         {
             VerifyRoutingIsRegistered(builder);
 
+            var routeOptions = builder.ApplicationServices.GetRequiredService<IOptions<RouteOptions>>();
+            EndpointDataSource middlewareEndpointDataSource;
+
             if (configure != null)
             {
-                var dataSourceBuilder = (DefaultEndpointDataSourceBuilder)builder.ApplicationServices.GetRequiredService<EndpointDataSourceBuilder>();
-                dataSourceBuilder.ApplicationBuilder = builder;
-                configure(dataSourceBuilder);
+                var endpointRouteBuilder = builder.ApplicationServices.GetRequiredService<IEndpointRouteBuilder>();
+                if (endpointRouteBuilder is DefaultEndpointRouteBuilder defaultEndpointRouteBuilder)
+                {
+                    defaultEndpointRouteBuilder.ApplicationBuilder = builder;
```

Comment:
?

Reply:
Yeah it is a little wierd. ApplicationBuilder isn't settable on the interface.

Reply:
The cleaner way to do this would be to DI the application builder factory. A cut a corner building this originally because it was for a demo.